### DISCUSSION TITLE
Fail with error 'SafeMath: subtraction overflow' & SafeMath error on sending tokens via contract

### DISCUSSION THREAD

**Discussion by Leptron-Technologies** [PavanAnanthSharma/Aave-FlashLoan-using-solidity#7] [Q&A] [answered]
The transaction runs cleanly, but no transferred tokens are displayed...

![83c36fe5f48b805ef7826062b366455ac3a6d366](https://user-images.githubusercontent.com/106593836/177317198-4e8f1d59-7e7c-45ff-bec4-f69a0d4a1dff.jpg)

**Comment by PavanAnanthSharma** [comment, marked as answer]
Assuming that the address calling ContractA hasn't set an allowance for the contract.
>Requirements:
* The caller must have an allowance for the sender's tokens of at least amount:: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v2.3.0/contracts/token/ERC20/ERC20.sol
* ``` /**
     * @dev See `IERC20.transferFrom`.
     *
     * Emits an `Approval` event indicating the updated allowance. This is not
     * required by the EIP. See the note at the beginning of `ERC20`;
     *
     * Requirements:
     * - `sender` and `recipient` cannot be the zero address.
     * - `sender` must have a balance of at least `value`.
     * - The caller must have an allowance for `sender`'s tokens of at least
     * `amount`.
     */
    function transferFrom(address sender, address recipient, uint256 amount) public returns (bool) {
        _transfer(sender, recipient, amount);
        _approve(sender, msg.sender, _allowances[sender][msg.sender].sub(amount));
        return true;
    }```
1. The caller of Contract A approves an allowance for the Token
2. The caller of Contract A calls the pay function and which calls transferFrom to transfer from the user to the recipient within the previously set allowance.

* If you are creating ERC20 tokens you may want to look at the OpenZeppelin Contracts implementation to see if this meets your needs. See the documentation for details: https://docs.openzeppelin.com/contracts/2.x/tokens#ERC20

* If you are transferring a variety of ERC20 tokens you may want to consider using SafeERC20 wrapper to make these calls: https://docs.openzeppelin.com/contracts/2.x/api/token/erc20#SafeERC20

* If you need an interface for ERC20 you may want to look at using IERC20: https://docs.openzeppelin.com/contracts/2.x/api/token/erc20#IERC20

* Alternatively to ERC20 you could look at creating ERC777 tokens (no need to do approve and transferFrom in two separate transactions). See the documentation for details: https://docs.openzeppelin.com/contracts/2.x/tokens#ERC777

* If you have questions on using OpenZeppelin you can ask in the Community Forum: https://forum.openzeppelin.com/

**Reply by Leptron-Technologies** [reply]
Thanks, sure will use OpenZeppelin and let you know if any changes are to be made!!!
